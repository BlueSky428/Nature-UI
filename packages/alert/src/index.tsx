import * as React from 'react';
import clx from 'clsx';
import { FiInfo, FiAlertCircle, FiAlertTriangle, FiCheckCircle } from 'react-icons/fi';

const SUBTLE_TEXT = 'text-gray-800';

export const ALERT_STATUSES = {
  error: {
    bg: 'bg-red-200',
    iconColor: 'text-red-600 mr-3',
    icon: FiAlertCircle,
    variant: {
      solid: 'bg-red-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  info: {
    bg: 'bg-blue-200',
    iconColor: 'text-blue-600 mr-3',
    icon: FiInfo,
    variant: {
      solid: 'bg-blue-600',
      subtle: SUBTLE_TEXT,
    },
  },
  success: {
    bg: 'bg-green-200',
    iconColor: 'text-green-600 mr-3',
    icon: FiCheckCircle,
    variant: {
      solid: 'bg-green-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  warning: {
    bg: 'bg-orange-200',
    iconColor: 'text-orange-600 mr-3',
    icon: FiAlertTriangle,
    variant: {
      solid: 'bg-orange-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
};

interface AlertOptions {
  /**
   * The status of the alert
   */
  status?: keyof typeof ALERT_STATUSES;
  /**
   * The variant of the alert style to use
   */
  variant?: 'solid' | 'subtle';

  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;

  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ElementType;
}

export type AlertProps = AlertOptions;

const BASE_STYLE = 'px-4 py-3 flex items-center';

const Alert: React.ForwardRefExoticComponent<AlertProps> = React.forwardRef(
  (props: AlertProps, ref: React.Ref<any>) => {
    const {
      className = '',
      status = 'success',
      children,
      variant = 'subtle',
      component: Component = 'div',
    } = props;

    // const Component = 'div';

    const VARIANT: string = ALERT_STATUSES[status].variant[variant];

    const componentClass = clx(BASE_STYLE, {
      [className]: className,
      [ALERT_STATUSES[status].bg]: status,
      [VARIANT]: variant,
    });

    const Icon = ALERT_STATUSES[status];
    const iconClasses = clx({
      [VARIANT]: variant,
      'mr-3': variant,
      [Icon.iconColor]: variant !== 'solid',
    });

    const IconComponent = Icon.icon;

    return (
      <Component className={componentClass} ref={ref}>
        {status && <IconComponent className={iconClasses} size={20} />}
        <span>{children ? children : 'This is an alert ðŸ™‚'}</span>
      </Component>
    );
  }
);

Alert.displayName = 'Alert';

export default Alert;
