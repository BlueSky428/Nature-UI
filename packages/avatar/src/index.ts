export * from './avatar';
export * from './avatarGroup';
