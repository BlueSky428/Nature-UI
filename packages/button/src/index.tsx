import * as React from 'react';
import clx from 'clsx';

interface ButtonProps {
  /**
   * The label of the button
   */
  'aria-label'?: string;
  /**
   * The text color of the button. Use a color key passed in theme.colors.
   */
  text?: string;
  /**
   * The background color of the button. Use a color key passed in theme.colors.
   */
  color?: string;
  /**
   * The variant of the button style to use.
   */
  variant?: 'outline' | 'ghost' | 'link' | 'solid' | 'none';
  /**
   * If true, the button will be disabled.
   */
  isDisabled?: boolean;
  /**
   * If true, the button will show a spinner.
   */
  isLoading?: boolean;
  /**
   * The label to show in the button when isLoading is true. If no text is passed, it only shows the spinner
   */
  loadingText?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'none';
  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;
  /**
   * The lightness or how dark you want the background color to be
   */
  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   * defaults to "button"
   */
  component?: React.ElementType;
  id?: string;
  /**
   *  when the href prop is set it changes the tag name to a
   */
  href?: string;
  onClick?: () => any;
}

const darken = (value: string): string => {
  const splitStr = value.split('-');
  const dept = Number(splitStr[1]);
  const color = splitStr[0];

  if (dept >= 800) {
    return `${color}-${dept - 200}`;
  }

  return `${color}-${dept + 200}`;
};

const lighten = (value: string): string => {
  const color = value.split('-')[0];

  return `${color}-${100}`;
};

const Button = React.forwardRef((props: ButtonProps, ref: React.Ref<any>) => {
  const {
    component: Component = 'button',
    variant = 'solid',
    color = 'teal-500',
    text = 'white',
    size = 'md',
    children,
    ...rest
  } = props;
  const DEFAULT_CLASS = 'block focus:shadow-outline focus:outline-none rounded font-semibold flex';
  const STYLES = {
    solid: `bg-${color} text-${text} hover:bg-${darken(color)}`,
    outline: `bg-transparent hover:bg-${lighten(
      text
    )} text-${text} border border-${text} focus:border-transparent`,
    ghost: `hover:bg-${lighten(text)} text-${text}`,
    link: `hover:underline text-${text}`,
    md: 'px-4 py-2',
  };

  let BTNClass: string;

  if (variant === 'none') {
    BTNClass = '';
  } else {
    BTNClass = clx(DEFAULT_CLASS, {
      [STYLES[size]]: size && variant !== 'link',
      [STYLES[variant]]: variant,
    });
  }

  return (
    <Component className={BTNClass} ref={ref} {...rest}>
      {children ? children : 'Button'}
    </Component>
  );
});

export default Button;
