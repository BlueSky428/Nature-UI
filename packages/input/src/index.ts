export * from './input';
export * from './input-addon';
