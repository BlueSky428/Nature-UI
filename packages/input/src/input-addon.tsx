/** @jsx jsx */
import { nature, PropsOf, clsx, jsx } from '@nature-ui/system';
import { __DEV__ } from '@nature-ui/utils';
import * as React from 'react';

type Placement = 'left' | 'right';
const SpanTag = nature('span');

const placements = {
  left: {
    marginRight: '-1px',
    borderRightRadius: 0,
    borderRightColor: 'transparent',
  },
  right: {
    marginLeft: '0px',
    borderLeftRadius: 0,
    borderLeftColor: 'transparent',
  },
};

/**
 * StyledAddon
 *
 * Wrapper element around the InputAddon component
 */
const StyledAddon = (props: PropsOf<typeof SpanTag>) => {
  const { className = '', ...rest } = props;

  const _className = clsx('flex w-auto items-center whitespace-no-wrap', {
    [className]: className,
  });

  return <SpanTag className={_className} {...rest} />;
};

export type InputAddonProps = PropsOf<typeof StyledAddon> & {
  placement?: Placement;
};

/**
 * InputAddon
 *
 * Element to append or prepend to an input
 */
export const InputAddon = React.forwardRef(
  (props: InputAddonProps, ref: React.Ref<any>) => {
    const { placement = 'left', children, className = '', ...rest } = props;
    const placementStyles = placements[placement] ?? {};
    // const _isString = typeof children === 'string';
    const _className = clsx(
      className,
      'px-4 bg-gray-100 border border-gray-200',
    );

    return (
      <StyledAddon
        ref={ref}
        css={placementStyles}
        {...rest}
        className={_className}
      >
        {children}
      </StyledAddon>
    );
  },
);

if (__DEV__) {
  InputAddon.displayName = 'InputAddon';
}

/**
 * InputLeftAddon
 *
 * Element to append to the left of an input
 */
export const InputLeftAddon = React.forwardRef(
  (props: InputAddonProps, ref: React.Ref<any>) => {
    const { className = '', ...rest } = props;
    const _className = clsx(className, 'rounded-l');

    return <InputAddon ref={ref} {...rest} className={_className} />;
  },
);

if (__DEV__) {
  InputLeftAddon.displayName = 'InputLeftAddon';
}

// @ts-ignore
InputLeftAddon.__hidden = 'InputLeftAddon';

/**
 * InputRightAddon
 *
 * Element to append to the right of an input
 */
export const InputRightAddon = React.forwardRef(
  (props: InputAddonProps, ref: React.Ref<any>) => {
    const { className = '', ...rest } = props;
    const _className = clsx(className, 'rounded-r');

    return (
      <InputAddon
        ref={ref}
        placement='right'
        className={_className}
        {...rest}
      />
    );
  },
);

if (__DEV__) {
  InputRightAddon.displayName = 'InputRightAddon';
}

// @ts-ignore
InputRightAddon.__hidden = 'InputRightAddon';
