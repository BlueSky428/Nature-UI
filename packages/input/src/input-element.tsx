/** @jsx jsx */
import { nature, clsx, jsx, PropsOf } from '@nature-ui/system';
import { __DEV__ } from '@nature-ui/utils';
import * as React from 'react';

const DivTag = nature('div');

export type InputElementProps = PropsOf<typeof DivTag> & {
  placement?: 'left' | 'right';
};

const StyledElement = (props: PropsOf<typeof DivTag>) => {
  const { className = '', ...rest } = props;

  const _className = clsx(`flex items-center justify-center absolute top-0`, {
    [className]: className,
  });

  return <DivTag className={_className} css={{ zIndex: 2 }} {...rest} />;
};

const InputElement = React.forwardRef(
  (props: InputElementProps, ref: React.Ref<any>) => {
    const { placement = 'left', ...rest } = props;

    const placementProp = { [placement]: '0' };

    return <StyledElement ref={ref} {...placementProp} {...rest} />;
  }
);

// @ts-ignore
InputElement.__hidden = 'InputElement';

if (__DEV__) {
  InputElement.displayName = 'InputElement';
}

export const InputLeftElement = React.forwardRef(
  (props: InputElementProps, ref: React.Ref<any>) => {
    const { className, ...rest } = props;
    const _className = clsx('chakra-input__left-element', className);

    return (
      <InputElement
        ref={ref}
        placement='left'
        className={_className}
        {...rest}
      />
    );
  }
);

// @ts-ignore
InputLeftElement.__hidden = 'InputLeftElement';

if (__DEV__) {
  InputLeftElement.displayName = 'InputLeftElement';
}

export const InputRightElement = React.forwardRef(
  (props: InputElementProps, ref: React.Ref<any>) => {
    const { className, ...rest } = props;
    const _className = clsx('chakra-input__right-element', className);

    return (
      <InputElement
        ref={ref}
        placement='right'
        className={_className}
        {...rest}
      />
    );
  }
);

// @ts-ignore
InputRightElement.__hidden = 'InputRightElement';

if (__DEV__) {
  InputRightElement.displayName = 'InputRightElement';
}
