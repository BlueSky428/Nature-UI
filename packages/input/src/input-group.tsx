/** @jsx jsx */
import { jsx, nature, PropsOf, clsx } from '@nature-ui/system';
import { __DEV__, getValidChildren } from '@nature-ui/utils';
import * as React from 'react';

const DivTag = nature('div');

export type InputGroupProps = PropsOf<typeof DivTag> & {
  size?: 'xs' | 'sm' | 'md' | 'lg' | number;
};

const _SIZES = {
  sm: '2rem',
  md: '2.5rem',
  lg: '3rem',
};

export const InputGroup = React.forwardRef(
  (props: InputGroupProps, ref: React.Ref<any>) => {
    const { children = '', size = 'md', className, ...rest } = props;

    const _className = clsx('flex', className);

    const stylesRef = React.useRef<InputGroupProps>({});

    const validChildren = getValidChildren(children);

    const __size = typeof size === 'string' ? _SIZES[size] : `${size}px`;

    validChildren.forEach((child: any) => {
      if (!__size) return;

      if (child.type.__hidden === 'InputLeftElement') {
        stylesRef.current.css = {
          paddingLeft: size,
        };
      }

      if (child.type.__hidden === 'InputRightElement') {
        // stylesRef.current.paddingRight = __size
      }

      if (child.type.__hidden === 'InputRightAddon') {
        (stylesRef as any).current.borderRightRadius = 0;
      }

      if (child.type.__hidden === 'InputLeftAddon') {
        (stylesRef as any).current.borderLeftRadius = 0;
      }
    });

    const clones = validChildren.map((child: any) => {
      return child.type.__hidden !== 'Input'
        ? React.cloneElement(child, {})
        : React.cloneElement(child, {
            css: {
              borderTopRightRadius: (stylesRef as any).current
                ?.borderRightRadius,
              borderBottomRightRadius: (stylesRef as any).current
                ?.borderRightRadius,
              borderTopLeftRadius: (stylesRef as any).current?.borderLeftRadius,
              borderBottomLeftRadius: (stylesRef as any).current
                ?.borderLeftRadius,
            },
          });
    });

    return (
      <nature.div className={_className} ref={ref} {...rest}>
        {clones}
      </nature.div>
    );
  }
);
