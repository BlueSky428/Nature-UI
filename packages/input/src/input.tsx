/** ** */
import { FormControlOptions } from '@nature-ui/form-control';
import {
  clsx,
  css,
  forwardRef,
  HTMLNatureProps,
  nature,
} from '@nature-ui/system';
import { __DEV__ } from '@nature-ui/utils';
import * as React from 'react';
import { InputLeftAddon, InputRightAddon } from './input-addon';

interface InputOptions {
  /**
   * The border color when the input is focused. Use color keys in `theme.colors`
   * @example
   * focusBorderColor = "blue-600"
   */
  focusBorderColor?: string;
  /**
   * The border color when the input is invalid. Use color keys in `theme.colors`
   * @example
   * errorBorderColor = "red-500"
   */
  errorBorderColor?: string;
  /**
   * If `true`, the input element will span the full width of it's parent
   */
  isFullWidth?: boolean;
  /**
   * Left addon: It can be an icon or any form of JSX element
   * and would be positioned the left.
   */
  addonLeft?: React.ReactNode;
  /**
   * Right addon: It can be an icon or any form of JSX element
   * and would be positioned the right.
   */
  addonRight?: React.ReactNode;
}

type Omitted = 'disabled' | 'required' | 'readOnly' | 'size';

const _SIZES = {
  sm: {
    height: '2rem',
    fontSize: '0.875rem',
  },
  md: {
    height: '2.5rem',
    fontSize: '1rem',
  },
  lg: {
    height: '3rem',
    fontSize: '1.125rem',
  },
};

const _PADDING = {
  sm: {
    paddingLeft: '0.75rem',
    paddingRight: '0.75rem',
  },
  md: {
    paddingLeft: '1rem',
    paddingRight: '1rem',
  },
  lg: {
    paddingLeft: '1rem',
    paddingRight: '1rem',
  },
};

export interface InputProps
  extends Omit<HTMLNatureProps<'input'>, Omitted>,
    InputOptions,
    FormControlOptions {
  size?: keyof typeof _SIZES | number;
  variant?: 'outline' | 'filled' | 'flushed' | 'unstyled';
}

/**
 * Input - Theming
 *
 * To style the input globally, change the styles in
 * `theme.components.Input`
 */
export const Input = forwardRef<InputProps, 'input'>((props, ref) => {
  const {
    className = '',
    size = 'md',
    isInvalid,
    isReadOnly,
    isDisabled,
    variant = 'outline',
    addonLeft,
    addonRight,
    focusBorderColor,
    errorBorderColor,
    ...rest
  } = props;

  const _addon = addonLeft || addonRight;

  const _padding = typeof size === 'string' && css(_PADDING[size]);
  const _css = typeof size === 'string' && css(_SIZES[size]);

  const _height =
    typeof size === 'number' &&
    css`
      height: ${size}px;
    `;

  const _border = css`
    &:focus {
      border-width: 2px;
    }
  `;

  const _focusBorderColor = focusBorderColor || 'blue-600';
  const _errorBorderColor = errorBorderColor || 'red-500';
  const _invalid = `focus:border-${_errorBorderColor} border-${_errorBorderColor} border-2`;

  const _outline = variant === 'outline';
  const _filled = variant === 'filled';
  const _flushed = variant === 'flushed';
  const _unstyled = variant === 'unstyled';

  const _className = clsx(
    'w-full outline-none transition-all duration-150',
    {
      [_invalid]: isInvalid,
      border: !isInvalid && _outline,
      [_border]: !isReadOnly && (_outline || _filled),
      'cursor-not-allowed opacity-50': isDisabled,
      'border-solid border-gray-200': _outline,
      [`focus:border-${_focusBorderColor}`]:
        !isInvalid && (_outline || _filled),
      'hover:bg-gray-300 bg-gray-200 focus:bg-transparent': _filled,
      [`${String(_padding)} rounded`]: _filled || _outline,
      [`border-b-2 border-gray-200 focus:border-${_focusBorderColor}`]:
        _flushed,
      [(String(_height), String(_css))]: !_unstyled,
    },
    className,
  );

  const _withAddon = clsx(
    {
      'rounded-l-none': addonLeft,
      'rounded-r-none': addonRight,
    },
    _className,
  );

  if (_addon) {
    return (
      <nature.span className='flex'>
        {addonLeft && (
          <InputLeftAddon className={String(_css)}>{addonLeft}</InputLeftAddon>
        )}
        <nature.input ref={ref} className={_withAddon} {...rest} />
        {addonRight && (
          <InputRightAddon className={String(_css)}>
            {addonRight}
          </InputRightAddon>
        )}
      </nature.span>
    );
  }

  return <nature.input ref={ref} className={_className} {...rest} />;
});

if (__DEV__) {
  Input.displayName = 'Input';
}

// @ts-ignore
Input.__hidden = 'Input';
