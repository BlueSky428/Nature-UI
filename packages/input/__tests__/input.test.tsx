import { render } from '@nature-ui/test-utils';
import { Input } from '../src';

describe('@nature-ui/input', () => {
  test('Input renders correctly', () => {
    const { asFragment } = render(<Input />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Input addons render correctly', () => {
    const { asFragment } = render(
      <Input addonLeft='https://' addonRight='.com' />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('Elements inside input render correctly', () => {
    const { asFragment } = render(
      <Input addonLeft={<span>Hello</span>} addonRight={<span>Hello</span>} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('Invalid input renders correctly', () => {
    const { getByTestId, container } = render(
      <Input isInvalid data-testid='input' />,
    );
    const input = getByTestId('input');
    console.log(container.innerHTML);

    expect(input).toHaveAttribute('aria-invalid', 'true');
  });

  test('Disabled input renders correctly', () => {
    const { getByTestId } = render(<Input isDisabled data-testid='input' />);
    const input = getByTestId('input');

    expect(input).toHaveAttribute('disabled');
  });

  test('Readonly input renders correctly', () => {
    const { getByTestId } = render(<Input isReadOnly data-testid='input' />);
    const input = getByTestId('input');

    expect(input).toHaveAttribute('aria-readonly', 'true');
  });
});
