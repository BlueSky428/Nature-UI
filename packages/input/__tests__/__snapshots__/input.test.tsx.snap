// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nature-ui/input Elements inside input render correctly 1`] = `
<DocumentFragment>
  <span
    class="flex"
  >
    <span
      class="flex w-auto items-center whitespace-no-wrap css-1ri2j3w rounded-l px-4 bg-gray-100 border border-gray-400 css-1btbaz3"
    >
      <span>
        Hello
      </span>
    </span>
    <input
      class="rounded-l-none rounded-r-none w-full outline-none transition-all duration-150 border css-nhqk3t border-solid border-gray-400 focus:border-blue-600 css-q6n4wc rounded css-1ri2j3w"
    />
    <span
      class="flex w-auto items-center whitespace-no-wrap css-1ri2j3w rounded-r px-4 bg-gray-100 border border-gray-400 css-hsw7ky"
    >
      <span>
        Hello
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`@nature-ui/input Input addons render correctly 1`] = `
<DocumentFragment>
  <span
    class="flex"
  >
    <span
      class="flex w-auto items-center whitespace-no-wrap css-1ri2j3w rounded-l px-4 bg-gray-100 border border-gray-400 css-1btbaz3"
    >
      https://
    </span>
    <input
      class="rounded-l-none rounded-r-none w-full outline-none transition-all duration-150 border css-nhqk3t border-solid border-gray-400 focus:border-blue-600 css-q6n4wc rounded css-1ri2j3w"
    />
    <span
      class="flex w-auto items-center whitespace-no-wrap css-1ri2j3w rounded-r px-4 bg-gray-100 border border-gray-400 css-hsw7ky"
    >
      .com
    </span>
  </span>
</DocumentFragment>
`;

exports[`@nature-ui/input Input renders correctly 1`] = `
<DocumentFragment>
  <input
    class="w-full outline-none transition-all duration-150 border css-nhqk3t border-solid border-gray-400 focus:border-blue-600 css-q6n4wc rounded css-1ri2j3w"
  />
</DocumentFragment>
`;
